### PR TITLE
Add a lit test for Xcode project generation

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
The test takes ~100s on my machine. 

Changes to CMake code tend to break Xcode project generation every other month or so. Although these issues are often fixed in a matter of days and are not as big of a deal for seasoned contributors, Xcode is the editor of choice for many, if not most, newcomers, and I believe this makes un interested in promising a functional invocation with the Xcode version used in CI.

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

File: validation-test/BuildSystem/generate_xcode.test [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# To save time, borrow these dependencies from the current build directories,
+# and test Xcode project generation only for 'swift'.
+
+# RUN: ln -s %swift_obj_root/../cmark-%target-os-%target-arch %t/B/cmark-%target-os-%target-arch
+# RUN: ln -s %swift_obj_root/../llvm-%target-os-%target-arch %t/B/llvm-%target-os-%target-arch
+# RUN: SWIFT_BUILD_ROOT=%t %swift_src_root/utils/build-script --cmake %cmake --swift-darwin-supported-archs="$(uname -m)" --build-subdir="B" --skip-build-cmark --build-llvm=0 --xcode
+
+# REQUIRES: standalone_build
+# REQUIRES: OS=macosx
+# RE QUIRES: long_test
```
Comment (xedin) [inline validation-test/BuildSystem/generate_xcode.test:18]: I am guessing that this space is not intentional?

Comment (AnthonyLatsis) [reply validation-test/BuildSystem/generate_xcode.test:18]: Oops, well noticed.

**Comment by AnthonyLatsis** [comment]
```
186.43s: Swift-validation(macosx-x86_64) :: BuildSystem/generate_xcode.test
```

File: utils/swift_build_support/swift_build_support/products/llvm.py [hunk 2]
```diff
@@ -216,20 +216,11 @@ def build(self, host_target):
...
-        if self.args.llvm_ninja_targets_for_cross_compile_hosts and \
-           self.is_cross_compile_target(host_target):
-            build_targets = (self.args.llvm_ninja_targets_for_cross_compile_hosts)
-        elif self.args.llvm_ninja_targets:
-            build_targets = (self.args.llvm_ninja_targets)
-
-        # indicating we don't want to build LLVM should
-        # override any custom ninja target we specified
         if not self.args._build_llvm:
-            build_targets = ['clean']
```
Comment (ahoppen) [inline utils/swift_build_support/swift_build_support/products/llvm.py:230]: I’m not sure why we used to add the `clean` target but it looks like we lost it. Is that intentional?

File: validation-test/BuildSystem/generate_xcode.test [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# (e.g. llvm-tblgen). This is why we still build them with '--skip-build' or
+# '--skip-build-llvm' set, and have an additional '--build-llvm' option to
+# actually skip building that bare minimum too.
+#
+# To save time, borrow these dependencies from the current build directories,
+# and test Xcode project generation only for 'swift'.
+
+# RUN: ln -s %swift_obj_root/../cmark-%target-os-%target-arch %t/B/cmark-%target-os-%target-arch
+# RUN: ln -s %swift_obj_root/../llvm-%target-os-%target-arch %t/B/llvm-%target-os-%target-arch
+# RUN: SWIFT_BUILD_ROOT=%t %swift_src_root/utils/build-script --cmake %cmake --swift-darwin-supported-archs="$(uname -m)" --build-subdir="B" --skip-build-cmark --build-llvm=0 --xcode
```
Comment (ahoppen) [inline validation-test/BuildSystem/generate_xcode.test:14]: Would a slightly more descriptive name for `--build-subdir` be good?

**Review by ahoppen — COMMENTED** [review]
Nice. 3 minutes is really good. Let’s add this.

File: validation-test/BuildSystem/generate_xcode.test (hunk 3 above)
Comment (AnthonyLatsis) [reply validation-test/BuildSystem/generate_xcode.test]: Is "Xcode" fine?

File: utils/swift_build_support/swift_build_support/products/llvm.py (hunk 2 above)
Comment (AnthonyLatsis) [reply utils/swift_build_support/swift_build_support/products/llvm.py:230]: Neither am I, but I doubt it makes a difference.

> Is that intentional?

Yes, I changed the logic slightly to not run any build commands for LLVM when `--build-llvm=0` (the "actually don’t build LLVM" option), so it no longer matters what `build_targets` is set to in this case. This behavior is consistent with how we handle other CMake products, and it allows the new test to not just succeed (a build command that was assembled for the Xcode generator will fail against a Ninja build directory due to [incompatible adjustment to the options](https://github.com/apple/swift/blob/7eac19998490f2912937e1ef271b4c43f8c12164/utils/swift_build_support/swift_build_support/products/cmake_product.py#L76-L87), but also run correctly (we don’t want to run any build commands against the symlinked LLVM build directory).

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
Hm, it seems like none of the macOS PR testing jobs run long tests. I can only see branch jobs running them. Do we keep the requirement?

File: validation-test/BuildSystem/generate_xcode.test (hunk 3 above)
Comment (ahoppen) [reply validation-test/BuildSystem/generate_xcode.test]: Yes, that’s good. `B` is just super ambiguous 😉

**Comment by ahoppen** [comment]
Let’s make it a non-long test then, I’d say.

File: utils/swift_build_support/swift_build_support/products/llvm.py (hunk 2 above)
Comment (ahoppen) [reply utils/swift_build_support/swift_build_support/products/llvm.py:230]: Sounds reasonable to me 👍🏽

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
When you @mention a GitHub username anywhere
in the context of an issue or pull request, that person is notified and
subscribed to future updates.

**Comment by AnthonyLatsis** [comment]
@shahmishal ping

**Review by shahmishal — APPROVED** [review]

**Comment by shahmishal** [comment]
@swift-ci smoke test

**Review by ahoppen — APPROVED** [review]

**Review by xedin — COMMENTED** [review]
LGTM!

**Pull request merged**